### PR TITLE
Allow converter.optional to take a Converter such as converter.pipe as its argument

### LINKED ISSUE #1348: Error occurs when using converters.optional and converters.pipe together
23.2.0 is OK.
```Python
Python 3.10.13 (main, Dec  6 2023, 12:01:00) [GCC 13.2.1 20230801] on linux
>>> import attr
>>> attr.__version__
'23.2.0'
>>> @attr.define
... class A:
...     abc: int | None = attr.field(converter=attr.converters.optional(attr.converters.pipe(str, int)), default=None)
...     
>>> A()
A(abc=None)
>>> A('1')
A(abc=1)
```

after 24.1.0, error occurs:
```Python
Python 3.10.13 (main, Dec  6 2023, 12:01:00) [GCC 13.2.1 20230801] on linux
>>> import attr
>>> attr.__version__
'24.2.0'
>>> @attr.define
... class A:
...     abc: int | None = attr.field(converter=attr.converters.optional(attr.converters.pipe(str, int)), default=None)
...     
>>> A()
A(abc=None)
>>> A('1')
Traceback (most recent call last):
  File "/opt/pycharm-professional/plugins/python-ce/helpers/pydev/pydevconsole.py", line 364, in runcode
    coro = func()
  File "<input>", line 1, in <module>
  File "<attrs generated init __main__.A>", line 3, in __init__
  File "/home/mio/Work/venv/calliper310N/lib/python3.10/site-packages/attr/converters.py", line 40, in optional_converter
    return converter(val)
TypeError: Converter.__init__.<locals>.<lambda>() missing 2 required positional arguments: 'instance' and 'field'

```

### PR DESCRIPTION
**Pull request by filbranden**
# Summary

Allow `converter.optional` to take a Converter such as `converter.pipe` as its argument.

Fixes #1348

# Pull Request Check List

- [x] Do **not** open pull requests from your `main` branch – **use a separate branch**!
  - There's a ton of footguns waiting if you don't heed this warning. You can still go back to your project, create a branch from your main branch, push it, and open the pull request from the new branch.
  - This is not a pre-requisite for your pull request to be accepted, but **you have been warned**.
- [x] Added **tests** for changed code.
  Our CI fails if coverage is not 100%.
- (N/A) New features have been added to our [Hypothesis testing strategy](https://github.com/python-attrs/attrs/blob/main/tests/strategies.py).
- (N/A) Changes or additions to public APIs are reflected in our type stubs (files ending in ``.pyi``).
- (N/A) Updated **documentation** for changed code.
- [x] Documentation in `.rst` and `.md` files is written using [semantic newlines](https://rhodesmill.org/brandon/2012/one-sentence-per-line/).
- [x] Changes (and possible deprecations) have news fragments in [`changelog.d`](https://github.com/python-attrs/attrs/blob/main/changelog.d).
- [x] Consider granting [push permissions to the PR branch](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork), so maintainers can fix minor issues themselves without pestering you.

### CODE REVIEW

**Review by Hnasar — CHANGES_REQUESTED** [review]
Thanks for starting on a fix!

File: src/attr/converters.py [hunk 1]
```diff
@@ -33,10 +33,25 @@ def optional(converter):
...
-    def optional_converter(val):
-        if val is None:
-            return None
-        return converter(val)
+    if isinstance(converter, Converter):
+
+        def optional_converter(val, inst, field):
+            if val is None:
+                return None
+            return converter(val, inst, field)
```
[the remark this answers is not in the document]
Comment (filbranden) [reply src/attr/converters.py:41]: Actually a `Converter` instance always takes three arguments: https://github.com/python-attrs/attrs/blob/13e9a6aaeac156bf792878eceaba10321728fa36/src/attr/_make.py#L2712-L2725

The callable it takes as an argument can take one, two or three, but the `Converter` class normalizes it to always take three arguments

So I believe this is correct

I think the typing information might need an update, to accept `Callable[[Any], Any] | Converter`  but I think that's a separate issue so maybe we should follow up on that on a separate PR?

File: tests/test_converters.py [hunk 2]
```diff
@@ -272,6 +280,46 @@ class C:
...
+
+        @attr.s
+        class C:
+            x = attrib(
+                converter=optional(pipe(str, Converter(to_bool), bool)),
+                default=None,
+            )
+
+        c1 = C()
+        assert None is c1.x
```
Comment (hynek) [inline tests/test_converters.py:317]:
```suggestion

        assert None is c1.x
```

File: tests/test_converters.py [hunk 3]
```diff
@@ -272,6 +280,46 @@ class C:
...
+
+        c2 = C("True")
+        assert True is c2.x
```
Comment (hynek) [inline tests/test_converters.py:320]:
```suggestion

        assert True is c2.x
```

**Review by hynek — APPROVED** [review]
Thanks for sorting it out yourself y'all! Feel free to open a PR re the typing issue you mentioned.

**Comment by filbranden** [comment]
Thanks for merging @hynek !

I have a follow up on typing on #1373 but looks like that's still failing CI, looks like mypy on 3.10+ breaks with the error in the comments (even though `tox -e py310-mypy` itself doesn't fail), I'll take another look later see if I can figure out why it's breaking.

I also posted #1374 to fix another issue I uncovered in converters, in a situation where they evaluate to False. Long story short, but doing an explicit comparison to None fixes that and I believe it makes sense here.

Cheers!

**Pull request merged**